packages/eslint-plugin/src/rules/base-to-string.ts

```diff
@@ -0,0 +1,130 @@
+import {
+  TSESTree,
+  AST_NODE_TYPES,
+} from '@typescript-eslint/experimental-utils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+enum Usefulness {
+  Always,
+  Never = 'will',
+  Sometimes = 'may',
+}
+
+export default util.createRule({
+  name: 'base-to-string',
+  meta: {
+    docs: {
+      description:
+        'Requires that `.toString()` is only called on objects which provide useful information when stringified',
+      category: 'Best Practices',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    messages: {
+      baseToString:
+        "'{{name}} {{certainty}} evaluate to '[Object object]' when stringified.",
+    },
+    schema: [],
+    type: 'suggestion',
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const typeChecker = parserServices.program.getTypeChecker();
+
+    function checkExpression(node: TSESTree.Expression, type?: ts.Type): void {
+      if (node.type === AST_NODE_TYPES.Literal) {
+        return;
+      }
+
+      const certainty = collectToStringCertainty(
+        type ??
+          typeChecker.getTypeAtLocation(
+            parserServices.esTreeNodeToTSNodeMap.get(node),
+          ),
+      );
+      if (certainty === Usefulness.Always) {
+        return;
+      }
+
+      context.report({
+        data: {
+          certainty,
+          name: context.getSourceCode().getText(node),
+        },
+        messageId: 'baseToString',
+        node,
+      });
+    }
+
+    function collectToStringCertainty(type: ts.Type): Usefulness {
+      const toString = typeChecker.getPropertyOfType(type, 'toString');
+      if (toString === undefined || toString.declarations.length === 0) {
+        return Usefulness.Always;
+      }
+
+      if (
+        toString.declarations.every(
+          ({ parent }) =>
+            !ts.isInterfaceDeclaration(parent) || parent.name.text !== 'Object',
+        )
+      ) {
+        return Usefulness.Always;
+      }
+
+      if (!type.isUnion()) {
+        return Usefulness.Never;
+      }
+
+      for (const subType of type.types) {
+        if (collectToStringCertainty(subType) !== Usefulness.Never) {
+          return Usefulness.Sometimes;
+        }
+      }
+
+      return Usefulness.Never;
+    }
+
+    return {
+      BinaryExpression(node: TSESTree.BinaryExpression): void {
```

Comment:
unless I'm mistaken, you only want to check `+` binary expressions?
- bitwise operators (`|`, `&`, `^`, `<<`, `>>`)
- comparison operators (`>`, `>=`, `==`, `===`, ...)
- other math operators (`*`, `**`, `-`, `/`)
- misc operators (`in`)
```suggestion
      'BinaryExpression[operator = "+"]'(node: TSESTree.BinaryExpression): void {
```

Reply:
it would probably be good to check for `x += obj` as well (`'AssignmentExpression[operator = "+="]'`)

Reply:
(don't forget to add always passing tests for some of the binary operators and assignment operators that aren't checked, just for sanity checking)